flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java

```diff
@@ -201,6 +203,18 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
 		}
 	}
 
+	/**
+	 * Query unreleased partitions.
+	 *
+	 * @return collection of partitions which still occupy some resources locally on this task executor
+	 * and have not been released yet. The partition can be released either with {@link ResultPartitionWriter#fail(Throwable)}
+	 * or {@link ResultPartitionWriter#finish()} and then with {@link ResultPartitionWriter#close()} after fail or finish.
```

Comment:
It might bring some confusing here.
`ResultPartitionWriter` has two different states `FINISHED` and `RELEASED` separately. `ResultPartitionWriter#finish()` only indicates the whole data produced and enters `FINISHED` state, but the partition is not released yet. The release could only happen in three scenarios atm

- `ResultPartitionWriter#fail` when task is failed or canceled
- `NetworkEnvironment#releasePartitions` from JM RPC call
- `ResultPartitionManager#onConsumedPartition` when partition is consumed by consumer side

In addition, it might be better to put these comments above and make `@return` short.

Reply:
In this comment I was actually thinking more about how to describe which user calls of shuffle API will lead to the release. `ResultPartitionWriter#finish()` triggers end of production and will result eventually internally in `ResultPartitionManager#onConsumedPartition`. But I think you are right we should mention that the actual release can happen later e.g. when the consumption is done. I will rewrite the comment a bit.

In general, `Close` could also potentially release some resources associated with partition, like memory now, we just do not associate it with release state of partition now because of current implementation details.